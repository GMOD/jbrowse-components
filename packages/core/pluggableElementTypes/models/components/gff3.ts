import { Feature } from '@jbrowse/core/util'

const coreFields = new Set([
  'uniqueId',
  'refName',
  'source',
  'type',
  'start',
  'end',
  'strand',
  'parent',
  'parentId',
  'score',
  'subfeatures',
  'phase',
])

const retitle = {
  id: 'ID',
  name: 'Name',
  alias: 'Alias',
  parent: 'Parent',
  target: 'Target',
  gap: 'Gap',
  derives_from: 'Derives_from',
  note: 'Note',
  description: 'Note',
  dbxref: 'Dbxref',
  ontology_term: 'Ontology_term',
  is_circular: 'Is_circular',
} as { [key: string]: string }

function fmt(obj: unknown): string {
  if (Array.isArray(obj)) {
    return obj.map(o => fmt(o)).join(',')
  } else if (typeof obj === 'object') {
    return JSON.stringify(obj)
  } else {
    return `${obj}`
  }
}

function formatFeat(f: Feature, parentId?: string, parentRef?: string) {
  return [
    f.get('refName') || parentRef,
    f.get('source') || '.',
    f.get('type') || '.',
    f.get('start') + 1,
    f.get('end'),
    f.get('score') || '.',
    f.get('strand') === 1 ? '+' : f.get('strand') === -1 ? '-' : '.',
    f.get('phase') || '.',
    (parentId ? `Parent=${parentId};` : '') +
      f
        .tags()
        .filter(tag => !coreFields.has(tag))
        .map(tag => [tag, fmt(f.get(tag))])
        .filter(tag => !!tag[1])
        .map(tag => `${retitle[tag[0]] || tag[0]}=${tag[1]}`)
        .join(';'),
  ].join('\t')
}
export function formatMultiLevelFeat(
  f: Feature,
  parentId?: string,
  parentRef?: string,
): string {
  const fRef = parentRef || f.get('refName')
  const fId = f.get('id')
  const primary = formatFeat(f, parentId, fRef)
  const subs =
    f.get('subfeatures')?.map(sub => formatMultiLevelFeat(sub, fId, fRef)) || []
  return [primary, ...subs].join('\n')
}

export function stringifyGFF3(feats: Feature[]) {
  return ['##gff-version 3', ...feats.map(f => formatMultiLevelFeat(f))].join(
    '\n',
  )
}
