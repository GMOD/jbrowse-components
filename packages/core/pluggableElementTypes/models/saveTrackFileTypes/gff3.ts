import type { Feature } from '@jbrowse/core/util'

const coreFields = new Set([
  'uniqueId',
  'refName',
  'source',
  'type',
  'start',
  'end',
  'strand',
  'parent',
  'parentId',
  'score',
  'subfeatures',
  'phase',
])

const retitle = {
  id: 'ID',
  name: 'Name',
  alias: 'Alias',
  parent: 'Parent',
  target: 'Target',
  gap: 'Gap',
  derives_from: 'Derives_from',
  note: 'Note',
  description: 'Note',
  dbxref: 'Dbxref',
  ontology_term: 'Ontology_term',
  is_circular: 'Is_circular',
} as Record<string, string>

function fmt(obj: unknown): string {
  if (Array.isArray(obj)) {
    return obj.map(o => fmt(o)).join(',')
  } else if (typeof obj === 'object' && obj !== null) {
    return JSON.stringify(obj)
  } else {
    return String(obj)
  }
}

function formatAttributes(f: Feature, parentId?: string) {
  const attributes = []
  if (parentId) {
    attributes.push(`Parent=${parentId}`)
  }

  const tags = Object.keys(f.toJSON()).filter(tag => !coreFields.has(tag))

  for (const tag of tags) {
    const val = f.get(tag)
    if (val !== undefined && val !== null) {
      const formattedVal = fmt(val)
      if (formattedVal) {
        const key = retitle[tag] || tag
        attributes.push(`${key}=${formattedVal}`)
      }
    }
  }
  return attributes.join(';')
}

function formatFeat(f: Feature, parentId?: string, parentRef?: string) {
  const strand = f.get('strand')
  return [
    f.get('refName') || parentRef,
    f.get('source') || '.',
    f.get('type') || '.',
    f.get('start') + 1,
    f.get('end'),
    f.get('score') || '.',
    strand === 1 ? '+' : strand === -1 ? '-' : '.',
    f.get('phase') || '.',
    formatAttributes(f, parentId),
  ].join('\t')
}
export function formatMultiLevelFeat(
  feature: Feature,
  parentId?: string,
  parentRef?: string,
): string {
  const featureRefName = parentRef || feature.get('refName')
  const featureId = feature.get('id')
  const primary = formatFeat(feature, parentId, featureRefName)

  return [
    primary,
    ...(feature
      .get('subfeatures')
      ?.map(sub => formatMultiLevelFeat(sub, featureId, featureRefName)) || []),
  ].join('\n')
}

export function stringifyGFF3({ features }: { features: Feature[] }) {
  return [
    '##gff-version 3',
    ...features.map(f => formatMultiLevelFeat(f)),
  ].join('\n')
}
