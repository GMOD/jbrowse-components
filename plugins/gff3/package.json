{
  "name": "@jbrowse/plugin-gff3",
  "version": "1.0.0",
  "description": "JBrowse 2 gff3.",
  "keywords": [
    "jbrowse",
    "jbrowse2"
  ],
  "license": "Apache-2.0",
  "homepage": "https://jbrowse.org",
  "bugs": "https://github.com/GMOD/jbrowse-components/issues",
  "repository": {
    "type": "git",
    "url": "https://github.com/GMOD/jbrowse-components.git",
    "directory": "plugins/gff3"
  },
  "author": "JBrowse Team",
  "distMain": "dist/index.js",
  "srcMain": "src/index.ts",
  "main": "src/index.ts",
  "distModule": "dist/plugin-gff3.esm.js",
  "module": "",
  "files": [
    "dist",
    "src"
  ],
  "scripts": {
    "start": "tsdx watch --verbose --noClean",
    "build": "tsdx build",
    "test": "cd ../..; jest plugins/gff3 --passWithNoTests",
    "prepublishOnly": "yarn test",
    "prepack": "yarn build; yarn useDist",
    "postpack": "yarn useSrc",
    "useDist": "node ../../scripts/useDist.js",
    "useSrc": "node ../../scripts/useSrc.js"
  },
  "dependencies": {
    "@gmod/gff": "^1.1.2",
    "@gmod/tabix": "^1.4.6",
    "generic-filehandle": "^2.0.0"
  },
  "peerDependencies": {
    "@jbrowse/core": "^1.0.0",
    "@jbrowse/plugin-linear-genome-view": "^1.0.0",
    "@material-ui/core": "^4.9.13",
    "mobx-react": "^6.0.0",
    "mobx-state-tree": "3.14.1",
    "prop-types": "^15.0.0",
    "react": "^16.8.0",
    "rxjs": "^6.0.0"
  },
  "publishConfig": {
    "access": "public"
  },
  "gitHead": "32e27bccb4eb35ee2b3ba3ab004baf68dfcbf38d"
}
