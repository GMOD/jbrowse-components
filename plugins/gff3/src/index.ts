import AdapterType from '@jbrowse/core/pluggableElementTypes/AdapterType'
import PluginManager from '@jbrowse/core/PluginManager'
import Plugin from '@jbrowse/core/Plugin'
import {
  AdapterClass as Gff3TabixAdapterClass,
  configSchema as gff3TabixAdapterConfigSchema,
} from './Gff3TabixAdapter'

export default class extends Plugin {
  name = 'GFF3TabixPlugin'

  install(pluginManager: PluginManager) {
    pluginManager.addAdapterType(
      () =>
        new AdapterType({
          name: 'Gff3TabixAdapter',
          configSchema: gff3TabixAdapterConfigSchema,
          AdapterClass: Gff3TabixAdapterClass,
        }),
    )
  }
}
