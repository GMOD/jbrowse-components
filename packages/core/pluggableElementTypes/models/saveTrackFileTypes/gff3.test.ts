import SimpleFeature, { Feature } from '@jbrowse/core/util/simpleFeature'
import { stringifyGFF3 } from './gff3'

// Helper function to create a feature, ensuring a unique ID is set
// for the SimpleFeature instance while using data.id for GFF3 attributes.
function createFeature(data: Record<string, any>): SimpleFeature {
  if (!data.id) {
    throw new Error('Test feature data must have an id')
  }
  return new SimpleFeature({ id: `${data.id}-unique`, data })
}

// Helper function to check GFF3 output using snapshot testing.
function expectGff3Snapshot(features: Feature[]) {
  expect(stringifyGFF3({ features })).toMatchSnapshot()
}

describe('GFF3 export', () => {
  it('can export a simple feature', () => {
    const f = createFeature({
      id: 'test',
      refName: 'chr1',
      start: 100,
      end: 200,
      type: 'gene',
      name: 'gene1',
      source: 'test',
    })
    expectGff3Snapshot([f])
  })

  it('can export a feature with score and strand', () => {
    const f = createFeature({
      id: 'test2',
      refName: 'chr1',
      start: 100,
      end: 200,
      type: 'gene',
      score: 100,
      strand: -1,
    })
    expectGff3Snapshot([f])
  })

  it('can export a feature with a subfeature', () => {
    const f = createFeature({
      id: 'gene1',
      refName: 'chr1',
      start: 100,
      end: 900,
      type: 'gene',
      subfeatures: [
        {
          id: 'exon1',
          start: 100,
          end: 200,
          type: 'exon',
        },
        {
          id: 'exon2',
          start: 800,
          end: 900,
          type: 'exon',
        },
      ],
    })
    expectGff3Snapshot([f])
  })

  it('can export a feature with multiple levels of subfeatures', () => {
    const f = createFeature({
      id: 'gene1',
      refName: 'chr1',
      start: 100,
      end: 900,
      type: 'gene',
      subfeatures: [
        {
          id: 'mrna1',
          start: 100,
          end: 900,
          type: 'mRNA',
          subfeatures: [
            {
              id: 'exon1',
              start: 100,
              end: 200,
              type: 'exon',
            },
            {
              id: 'exon2',
              start: 800,
              end: 900,
              type: 'exon',
            },
          ],
        },
      ],
    })
    expectGff3Snapshot([f])
  })

  it('handles various attribute types', () => {
    const f = createFeature({
      id: 'test3',
      refName: 'chr1',
      start: 100,
      end: 200,
      type: 'tf_binding_site',
      note: 'some note',
      alias: ['a', 'b'],
      custom: { foo: 'bar' },
      empty_array: [],
      undef: undefined,
      null_val: null,
    })
    expectGff3Snapshot([f])
  })
})