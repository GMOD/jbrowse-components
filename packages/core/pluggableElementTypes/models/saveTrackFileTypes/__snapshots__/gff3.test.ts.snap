// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GFF3 export can export a feature with a subfeature 1`] = `
"##gff-version 3
chr1	.	gene	101	900	.	.	.	ID=gene1
chr1	.	exon	101	200	.	.	.	Parent=gene1;ID=exon1
chr1	.	exon	801	900	.	.	.	Parent=gene1;ID=exon2"
`;

exports[`GFF3 export can export a feature with multiple levels of subfeatures 1`] = `
"##gff-version 3
chr1	.	gene	101	900	.	.	.	ID=gene1
chr1	.	mRNA	101	900	.	.	.	Parent=gene1;ID=mrna1
chr1	.	exon	101	200	.	.	.	Parent=mrna1;ID=exon1
chr1	.	exon	801	900	.	.	.	Parent=mrna1;ID=exon2"
`;

exports[`GFF3 export can export a feature with score and strand 1`] = `
"##gff-version 3
chr1	.	gene	101	200	100	-	.	ID=test2"
`;

exports[`GFF3 export can export a simple feature 1`] = `
"##gff-version 3
chr1	test	gene	101	200	.	.	.	ID=test;Name=gene1"
`;

exports[`GFF3 export handles various attribute types 1`] = `
"##gff-version 3
chr1	.	tf_binding_site	101	200	.	.	.	ID=test3;Note=some note;Alias=a,b;custom={"foo":"bar"}"
`;
